src/core/evaluator.js

```diff
@@ -1513,6 +1542,7 @@ class PartialEvaluator {
     const localColorSpaceCache = new LocalColorSpaceCache();
     const localGStateCache = new LocalGStateCache();
     const localTilingPatternCache = new LocalTilingPatternCache();
+    const localPatternCache = new Map();
```

Comment:
Would it perhaps help to implement a more "proper" cache here, see the existing ones in https://github.com/mozilla/pdf.js/blob/master/src/core/image_utils.js, since that may help improve overall caching since you could then cache by e.g. actual `Ref`s instead?

(This may, or may not, help all that much for the particular data in question, but I wanted to ask.)

Reply:
Also, perhaps `localShadingPatternCache` instead to provide a slightly clearer distinction between this new cache and the existing one just above?

Reply:
I started out using the localTilingPatternCache, but found some issues with it and and was planning to file some bugs. The patterns can't be cached by name alone since they could be in different resources dictionaries with duplicate names (issue6737.pdf has an example, but with radial gradients). There also seemed to be a number of pdfs where the pattern/shading dictionary is inline and caching by ref can't be used. 

As for perf, I don't think caching by ref should speed it up much, the shading/pattern dict should be in the xref's cache map, so it should be a pretty fast lookup.

Reply:
> I started out using the localTilingPatternCache,

You probably would've needed to implement a new and *separate* cache, rather than simply re-using the existing `localTilingPatternCache` for the Shadings-case that this patch is about.

> but found some issues with it and and was planning to file some bugs.

What sort of bugs did you encounter?

> There also seemed to be a number of pdfs where the pattern/shading dictionary is inline and caching by ref can't be used.
>
> As for perf, I don't think caching by ref should speed it up much, the shading/pattern dict should be in the xref's cache map, so it should be a pretty fast lookup.

I wasn't really worried about the `Dict` lookups themselves, since as you say the data is being cached, but was mostly wondering if caching by `Ref` could possibly have improved the hit/miss ratio in the cache. It doesn't sound like that'd been the case here though :-)

Reply:
Tiling patterns and shadings can both be patterns, so I should have been able to re-use that cache. I filed #13780 to further explain the issue.